### ISSUE TITLE
vktrace: multi-thread API calls are re-ordered in the trace

### ISSUE THREAD

**Issue by justuswang** [LunarG/VulkanTools#473]
When tracing a mult-thread Vulkan application, I have found that some API-calls are re-ordered in the trace file because current vktrace is recording API calls in completion order instead of issue order.  Packet with smaller global_packet_index sometimes is not stored before a packet with larger global_packet_index.

The observation is:

Thread 1: (start create image#1 ... end) -> (start destroy image#1 .............. end) --------> (other API)
Thread 2: ---------------------------------------> (start create image#1 ... end) ------> (start destroy image#1 ... end)

There's a vkDestroyImage being called after image#1 (a VkImage handle which does not mean identical image in thread#1 and thread#2) was destroyed on a different thread.

vkDestroyImage in thread#1 is called before vkCreateImage being called in thread#2 but completed after vkCreateImage in thread#2 being completed.

vkDestroyImage in thread#1 takes longer time than vkCreatImage in thread#2 which makes it being saved to trace file as a trace packet after the trace packet for vkCreateImage in thread#1.

The issue happens when tracing a multi-thread Vulkan application, every API call will be saved to trace file in their completion order instead of issue order which will cause unexpected overlap.

Is this a bug of vktrace?  Does vktrace need to have a mutex lock in every API call?  e.g. Add vktrace_enter_critical_section at the beginning of an API call and vktrace_leave_critical_section at the end of the API call.

**Comment by davidlunarg** [comment]
Yes, this is a bug in vktrace. The api calls should be added to the trace file in the order they are made. A mutex would solve the issue, but might negatively affect performance. Another solution might be to break up the packets into a "call packet" and a "result packet", in which the call packet is added to the trace file before the trace layer calls down to the next layer, and the result packet would be added after the call to the next layer returns. We're currently looking at other changes to vktrace in which such a strategy might be used. It will be a while before we make those changes, so maybe we want to implement the mutex solutions in the meantime?

**Comment by justuswang** [comment]
> We're currently looking at other changes to vktrace in which such a strategy might be used. It will be a while before we make those changes, so maybe we want to implement the mutex solutions in the meantime?

Since those changes won't happen in a short time, shall I create a pull request to implement the mutex to fix it for now?  Though it might negatively affect performance.  I think it's better to have a trace file with correct order than have a better performance when tracing an app.  And I prefer to have the fix ASAP.

**Comment by davidlunarg** [comment]
> Since those changes won't happen in a short time, shall I create a pull request to implement the mutex to fix it for now?
Though it might negatively affect performance. I think it's better to have a trace file with correct order than have a better performance when tracing an app. And I prefer to have the fix ASAP.

I agree that the trace file should be in the correct order. Yes, please submit a pull request.

**Comment by karl-lunarg** [comment]
I don't understand the description.  For example:

> vkDestroyImage in thread#1 takes longer time than vkCreatImage in thread#2 which makes it being saved to trace file as a trace packet after the trace packet for vkCreateImage in thread#1.

Isn't this what you want?  The destroy packet in thread 1 comes after the create packet in thread 1.

Maybe it would help if you list the order you saw the packets in the trace file for the 4 operations (thread 1: create image, thread 1: destroy image, thread 2: create image, thread 2: destroy image).

**Comment by justuswang** [comment]
> Maybe it would help if you list the order you saw the packets in the trace file for the 4 operations (thread 1: create image, thread 1: destroy image, thread 2: create image, thread 2: destroy image).

Sure, here's the original order (API call's issue order) in the app:

1. thread 1: create image (image handle#1)
2. thread 1: destroy image (image handle#1)
3. thread 2: create image (image handle#1 which is re-used after destroy)
4. thread 2: destroy image (image handle#1)

and following is the recorded order (API call's completion order) in the trace file:

1. thread 1: create image (image handle#1)
2. thread 2: create image (image handle#1)
3. thread 2: destroy image (image handle#1)
4. thread 1: destroy image (image handle#1)

This is because "_thread 1: destroy image_" is issued before "_thread 2: create image_" but completed after "_thread 2: destroy image_" though the image may already being destroyed in GPU driver before "_2. thread 2: create image_" but takes a long time to be returned from GPU driver.

**Comment by justuswang** [comment]
> I agree that the trace file should be in the correct order. Yes, please submit a pull request.

Sure, https://github.com/LunarG/VulkanTools/pull/479 has been submitted to fix it for now.

**Comment by karl-lunarg** [comment]
OK, the part I missed was the handle reuse.  Thanks.
